src/library_syscall.js

```diff
@@ -1439,13 +1445,18 @@ var SyscallsLibrary = {
     return 0;
   },
   fd_fdstat_get: function(fd, pbuf) {
+#if SYSCALLS_REQUIRE_FILESYSTEM
     var stream = SYSCALLS.getStreamFromFD(fd);
     // All character devices are terminals (other things a Linux system would
     // assume is a character device, like the mouse, we have special APIs for).
     var type = stream.tty ? {{{ cDefine('__WASI_FILETYPE_CHARACTER_DEVICE') }}} :
                FS.isDir(stream.mode) ? {{{ cDefine('__WASI_FILETYPE_DIRECTORY') }}} :
                FS.isLink(stream.mode) ? {{{ cDefine('__WASI_FILETYPE_SYMBOLIC_LINK') }}} :
                {{{ cDefine('__WASI_FILETYPE_REGULAR_FILE') }}};
+#else
+    // hack to support printf in SYSCALLS_REQUIRE_FILESYSTEM=0
+    var type = fd == 1 || fd == 2 ? {{{ cDefine('__WASI_FILETYPE_CHARACTER_DEVICE') }}} : abort();
```

Comment:
when is this needed? That is, in what situation can `fd_fdstat_get` be called but `SYSCALLS_REQUIRE_FILESYSTEM` not be set?

Reply:
I am building ffmpeg without FS, and that was already working fine with 1.38.x (see https://github.com/emscripten-core/emscripten/issues/10374 for further details)

There is a similar fallback code in`fd_write` that makes fputs/fprintf+stderr usable when compiling without FS support, such a similar fallback code/hack was just missing here on this function.

(I could also be compiling ffmpeg without `HAVE_ISATTY` to workaround the issue, but i thought more convenient submitting this patch to prevent the exceptions :))

Reply:
I see, thanks. Looks good!